src/Host.cpp

```diff
@@ -217,59 +222,478 @@ void Host::saveModules(int sync)
+                struct zip_source* zipSource = zip_source_file(archive, filename_xml.toStdString().c_str(), 0, 0);
+                if (!zipSource) {
+                    zip_error_t* pError = zip_get_error(archive);
+                    postMessage(tr("[ ERROR ] - Failed to open file \"%1\" to place into module (archive)\n"
+                                               "file \"%2\",\n"
+                                               "error is: \"%3\".")
+                                .arg(filename_xml, entry.at(0), QString::fromUtf8(zip_error_strerror(pError))));
+                    zip_error_fini(pError);
+                    zip_discard(archive);
+                    archive = 0;
+                    // Try again for next module...
+                    continue;
+                }
+
+                //  Might want to include ZIP_FL_OVERWRITE flag to later version:
+                zip_int64_t index = zip_file_add(archive, QStringLiteral("%1.xml").arg(moduleName).toStdString().c_str(), zipSource, ZIP_FL_OVERWRITE|ZIP_FL_ENC_UTF_8);
+
+                // index is -1 on error:
+                if (index == -1) {
+                    zip_error_t* pError = zip_get_error(archive);
+                    postMessage(tr("[ ERROR ] - Failed to replace file \"%1\" in module (archive)\n"
+                                               "file \"%2\",\n"
+                                               "error is: \"%3\".")
+                                .arg(filename_xml, entry.at(0), QString::fromUtf8(zip_error_strerror(pError))));
+                    zip_error_fini(pError);
+                    zip_source_close(zipSource);
+                    zip_discard(archive);
+                    archive = 0;
+                    // Try again for next module...
+                    continue;
+                }
+
+                if (zip_close(archive) == -1) {
+                    zip_error_t* pError = zip_get_error(archive);
+                    postMessage(tr("[ ERROR ] - Failed to replace file \"%1\" in module (archive)\n"
+                                               "file \"%2\",\n"
+                                               "error is: \"%3\".")
+                                .arg(filename_xml, entry.at(0), QString::fromUtf8(zip_error_strerror(pError))));
+                    zip_error_fini(pError);
+                    zip_discard(archive);
+                } // End of error on closing archive (at the right time)
+            } // End of zipName not being empty
         }
-        QFile file_xml(filename_xml);
-        if (file_xml.open(QIODevice::WriteOnly)) {
-            XMLexport writer(this);
-            writer.writeModuleXML(&file_xml, it.key());
-            file_xml.close();
-
-            if (entry[1].toInt()) {
-                modulesToSync << it.key();
+#elif defined(LIBZIP_VERSION_MAJOR) && defined(LIBZIP_VERSION_MINOR) && (LIBZIP_VERSION_MAJOR == 0) && (LIBZIP_VERSION_MINOR >= 11)
```

Comment:
We can get rid of this, right? We upped the libzip version to 1.x: https://github.com/Mudlet/Mudlet/commit/3e83c5f5c4b169e4867ac9ad512e9cf090de5137

Reply:
Only in the CI builds, some __users'__ systems still have 0.1x versions around IIRC.

Reply:
Who?

Not the deployed Windows, macOS, or Linux downloads.

Reply:
People building on more mature systems or from sources on _other_ platforms!  For an example of the latter consider [Cygwin](https://cygwin.com/cgi-bin2/package-grep.cgi?grep=libzip&arch=x86_64) currently on 0.11.2 and remember that some LTS versions of some more familiar places like Debian OldStable "Jessie" which does not reach EOL until June next year and that uses 0.11.2; or Ubuntu "Trusty Tahr" 14.04 LTS (which is what originally made me look at this) uses 0.10.1 and that does not reach EOL until around the time the UK leaves Europe (😢) March/April 2019.

Whilst in some case it can be got around by compiling a newer version that is harder to do when working on a system which prefers to keep things packaged - it is usually a lot simpler when creating a Mudlet package if the other things needed are also available in existing packages and can be pulled in using whatever system the platform uses...

Reply:
I'm personally not going to go out of my way to make the code more complicated for someone who prefers to run old OS's while using new software (what a paradox), but I appreciate the effort you've put into this let's have it stay then.

Reply:
It will help to get Mudlet to compile on other OSs which are not so bleeding edge (mind you since Cygwin dropped support for Windows XP my development machine surface has dropped considerable.

I only have 2 PCs with a later Doze version - a 32-bit dual core Win7 {dual booting to Devuan "Jessie"} and a 64-bit quad core Win7 {triple booting to Debian "Stretch" and FreeBSD 11.1 - I have abandoned trying to install GNU/Hurd for the moment} to work on. 🙂